### ISSUE TITLE
Action proposal: rollback

### ISSUE THREAD

**Issue by jlubeck** [fastlane/fastlane#9749]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

I read on the docs that I needed to create an issue before making a pull request for an action. So here it is. And here is the scenario:

You pushed a build to production only to realize there is a serious crash and you need to submit a new build as soon as possible. This action allows you specify the commit hash and new build number that you need to submit a new build from old code. It also has an option to send a lane name that will be executed once the rollback is complete.

I have it written already, and has helped me bring my rollback speed from 12 minutes, dealing with manually typing git commands, changing build numbers and running specific lanes,  down to 2 minutes with a single command. 

Would this be something worth making a pull request?

Thanks!

**Comment by KrauseFx** [comment]
Hey that sounds awesome, this is a great candidate for a [fastlane plugin](https://docs.fastlane.tools/plugins/create-plugin/) instead of an action 👍  Let me know if you need any help get up and running 🚀

**Comment by jlubeck** [comment]
Oh, ok. By reading at the docs it sounded like an action more than a plugin.

Taken from here: https://docs.fastlane.tools/plugins/create-plugin/

Actions:
- _Generally usable for a big majority of developers (e.g. **basic interactions with git**)_. This action is mainly that. 
- _Solve pain points for mobile app developers (iOS and Android)._ I just though that removing the manual process of having to roll back to an old build during an emergency would be "pain solving" haha. I actually had the need (unfortunately) on Wednesday. And I initially came here trying to find something that did this. I couldn't find anything so we had to do it manually and it took a teammate hours to do it. After the emergency passed, I tried it myself and it took me just 12 minutes (I've been in the company longer, so I know some shortcuts that he didn't, but I wasn't there when they needed it). Anyway, long story short, I know have this action that will help my teammates here in the future in case they need it and thought someone else might benefit from it as well. 

Plugins:
_This is useful if you maintain your own library or web service, and want to make sure the fastlane plugin is always up to date._ This is not for a web service nor library, that's why I thought it was good for an action...

Any way, here is the action code in case you want to give it a second look and change your mind.

```ruby
module Fastlane
  module Actions
    class RollbackAction < Action
      def self.run(options)
        output_directory = options[:output_directory]
        git_url = options[:git_url]
        git_commit_hash = options[:git_commit_hash]
        new_build_num = options[:new_build_num]
        lane_name = options[:lane_name]

        UI.message("Cloning #{git_url}.")

        Actions.sh("git clone #{git_url} #{output_directory}")
        UI.success("Successfully cloned #{git_url}.")

        UI.message("Checking out specific commit #{git_commit_hash}.")

        Actions.sh("cd #{output_directory} && git reset --hard #{git_commit_hash}")
        UI.success("Successfully checked out #{git_commit_hash}.")

        UI.message("Updating build number to #{new_build_num}.")
        plist_files = Dir.glob("#{output_directory}/**/*Info*.plist")
        for plist in plist_files do
          Actions::UpdateInfoPlistAction.run(
            plist_path: plist,
            block: lambda { |plist|
              plist["CFBundleVersion"] = "#{new_build_num}"
            }
          )
        end

        if lane_name
        UI.message("Running lane #{lane_name}.")
        Actions.sh("cd #{output_directory} && bundle exec fastlane #{lane_name}")
        end
        UI.success("")
        UI.success("🎉 Rollback complete 🎉")
        if !lane_name
        UI.success("Change directory into '#{output_directory}' and run 'bundle exec fastlane [lane_name]' to get the desired build")
        end
      end

      #####################################################
      # @!group Documentation
      #####################################################

      def self.description
        "Rollback your code to a previous commit"
      end

      def self.details
        "You pushed a build to production only to realize there is a serious crash and you need to submit a new build as soon as possible. This action allows you specify the commit hash and new build number that you need to submit a new build from old code."
      end

      def self.available_options
        [
          FastlaneCore::ConfigItem.new(key: :output_directory,
            env_name: "OUTPUT_DIRECTORY",
            description: "Path for the rollback workspace",
            default_value: "rollback",
            verify_block: proc do |value|
            UI.user_error!("No output directory given, pass using `output_directory: 'directory'`") unless value and !value.empty? end
            ),
          FastlaneCore::ConfigItem.new(key: :git_url,
            env_name: "GIT_URL",
            description: "Repository clone URL",
            verify_block: proc do |value|
            UI.user_error!("No clone URL given, pass using `git_url: 'url'`") unless value and !value.empty? end
            ),
          FastlaneCore::ConfigItem.new(key: :git_commit_hash,
            env_name: "GIT_COMMIT_HASH",
            description: "Commit hash to roll back to",
            verify_block: proc do |value|
            UI.user_error!("No commit hash given, pass using `git_commit_hash: 'hash'`") unless value and !value.empty? end
            ),
          FastlaneCore::ConfigItem.new(key: :new_build_num,
            env_name: "NEW_BUILD_NUM",
            description: "New build number to use after the rollback",
            verify_block: proc do |value|
            UI.user_error!("No build number given, pass using `new_build_num: 'number'`") unless value and !value.empty? end
            ),
          FastlaneCore::ConfigItem.new(key: :lane_name,
            env_name: "LANE_NAME",
            description: "Lane name number to run after the rollback",
            optional:true
            )
        ]
      end

      def self.authors
        ["jlubeck"]
      end

      def self.is_supported?(platform)
        platform == :ios
      end
    end
  end
end
```


Thanks!!

**Comment by KrauseFx** [comment]
Hey @jlubeck, yeah, but since this use-case isn't too common and it's pretty complex, I think it would better fit a plugin 👍

**Comment by KrauseFx** [comment]
If many people use it, it can always become an official action 👍

**Comment by jlubeck** [comment]
Sounds good, I'll see how to make it into a Plugin then. Thanks!

**Comment by KrauseFx** [comment]
Thanks 👍
